src/dune/lib.ml

```diff
@@ -1118,80 +1123,78 @@ end = struct
                 match
-                  List.find_map choices
-                    ~f:(fun { required; forbidden; file } ->
-                      if
-                        Lib_name.Set.exists forbidden
-                          ~f:(available_internal db ~stack)
-                      then
-                        None
-                      else
-                        match
-                          let deps =
-                            Lib_name.Set.fold required ~init:[]
-                              ~f:(fun x acc -> (loc, x) :: acc)
-                          in
-                          resolve_simple_deps ~allow_private_deps db deps
-                            ~stack
-                        with
-                        | Ok ts -> Some (ts, file)
-                        | Error _ -> None)
+                  let deps =
+                    Lib_name.Set.fold required ~init:[] ~f:(fun x acc ->
+                        (loc, x) :: acc)
+                  in
+                  resolve_simple_deps ~allow_private_deps db deps ~stack
                 with
-                | Some (ts, file) -> (Ok ts, Ok file)
-                | None ->
-                  let e () = Error.no_solution_found_for_select ~loc in
-                  (e (), e ())
-              in
-              ( res
-              , { Resolved_select.src_fn; dst_fn = result_fn } :: acc_selects
-              )
-          in
-          let res =
-            match (res, acc_res) with
-            | Ok l, Ok acc -> Ok (List.rev_append l acc)
-            | (Error _ as res), _
-             |_, (Error _ as res) ->
-              res
-          in
-          (res, acc_selects))
-    in
-    let res =
-      match res with
-      | Ok l -> Ok (List.rev l)
-      | Error _ -> res
+                | Ok ts -> Some (ts, file)
+                | Error _ -> None)
+        with
+        | Some (ts, file) -> (Ok ts, Ok file)
+        | None ->
+          let e () = Error.no_solution_found_for_select ~loc in
+          (e (), e ())
+      in
+      (res, { Resolved_select.src_fn; dst_fn = result_fn })
     in
-    (res, resolved_selects)
-
-  let resolve_deps db deps ~allow_private_deps ~stack =
-    (* Compute transitive closure *)
-    let libs, selects =
-      match (deps : Lib_info.Deps.t) with
-      | Simple names ->
-        (resolve_simple_deps db names ~allow_private_deps ~stack, [])
-      | Complex names ->
-        resolve_complex_deps db names ~allow_private_deps ~stack
+    let res, resolved_selects, re_exports =
+      List.fold_left deps ~init:(Ok [], [], Ok [])
+        ~f:(fun (acc_res, acc_selects, acc_re_exports) dep ->
+          match (dep : Lib_dep.t) with
+          | Re_export (loc, name) ->
+            let acc_re_exports =
+              let* lib = resolve_dep db name ~allow_private_deps ~loc ~stack in
+              let+ acc_re_exports = acc_re_exports in
+              lib :: acc_re_exports
+            in
+            (acc_res, acc_selects, acc_re_exports)
```

Comment:
Why don't we add `lib` to `acc_res`? IIIUC, re-exported libraries are not part of `t.requires`. That seems odd to me, in particular it seems that they don't end up in `Compile.direct_requires`.

Reply:
I can do that, but I think we still need to maintain a separate list of re_exports. We need to walk that list to get a list of all re_exported libraries.

Reply:
Actually, they do end up in direct_requires. `resolve_user_deps` resolves all re_exported dependencies.

Reply:
Yeah, a separate re-exports list seems fine. Basically I was just finding it odd that `re_exports` wasn't included in `res` given that each re-exported dependency is also a plan dependency.

Reply:
For instance, dependencies coming from `select` forms are also included in `res`